pkg/policy/l4.go

```diff
@@ -783,32 +789,43 @@ const (
 	OriginatingTLS TLSDirection = "originating"
 )
 
+// getCerts reads certificates out of the PolicyContext, reading from k8s or local files depending on config
+// and puts the values into the relevant keys in the TLSContext. Note that if the returned TLSContext.FromFile is
+// `false`, then this will be read from Kubernetes.
 func (l4 *L4Filter) getCerts(policyCtx PolicyContext, tls *api.TLSContext, direction TLSDirection) (*TLSContext, error) {
 	if tls == nil {
 		return nil, nil
 	}
-	ca, public, private, err := policyCtx.GetTLSContext(tls)
+	ca, public, private, inlineSecrets, err := policyCtx.GetTLSContext(tls)
 	if err != nil {
 		log.WithError(err).Warningf("policy: Error getting %s TLS Context.", direction)
 		return nil, err
 	}
-	switch direction {
-	case TerminatingTLS:
-		if public == "" || private == "" {
-			return nil, fmt.Errorf("Terminating TLS context is missing certs.")
-		}
-	case OriginatingTLS:
-		if ca == "" {
-			return nil, fmt.Errorf("Originating TLS context is missing CA certs.")
+
+	// If the secret is not being included into NPDS inline, we're going to pass an SDS reference instead.
+	if inlineSecrets {
+		switch direction {
+		case TerminatingTLS:
+			if public == "" || private == "" {
+				return nil, fmt.Errorf("Terminating TLS context is missing certs.")
+			}
+		case OriginatingTLS:
+			if ca == "" {
+				return nil, fmt.Errorf("Originating TLS context is missing CA certs.")
+			}
+		default:
+			return nil, fmt.Errorf("invalid TLS direction: %s", direction)
 		}
-	default:
-		return nil, fmt.Errorf("invalid TLS direction: %s", direction)
+	} else {
+		log.Debug("Secret being read from Kubernetes", "secret", types.NamespacedName(*tls.Secret))
 	}
 
 	return &TLSContext{
 		TrustedCA:        ca,
 		CertificateChain: public,
 		PrivateKey:       private,
+		FromFile:         inlineSecrets,
+		Secret:           types.NamespacedName(*tls.Secret),
```

Comment:
question about a specific use-case:

does applying the Cilium specific xDS resource `NPDS` fail (result in a NACK) if the referenced Secret isn't available via SDS? (e.g. in case the k8s Secret isn't applied yet)

If this would fail, and the K8s secret (that then gets synced via SDS) is created afterwards - I'd assume that the user has to trigger a change on the policy to re-trigger a reconciliation of the Policy calculation (and re-application of the xDS resource NPDS).

(Therefore, my hope is that applying NPDS doesn't fail if the  secret isn't available via SDS yet. And above manual interaction isn't necessary at all.)

Reply:
No, it doesn't fail, because NPDS doesn't check SDS until _after_ the config is ACKed. (All the xDS services work this way, and NPDS is no different).

It's one of the big advantages of the SDS approach, actually, that you no longer need to do a policy regeneration to update a Secret.